### ISSUE TITLE
Long traceback hid missing term error

### ISSUE THREAD

**Issue by mrocklin** [numba/numba#2254]
Hey, so, I got a strange error that it took me a while to find the problem.  I have now resolved my issue, but I thought I'd submit an issue report just in case this case has some value to you all.

Generally numba compilation failed because, I think, a variable was defined within a for loop / conditional rather than top scope.  It is, I think, impossible for this variable not to be defined, but I wouldn't expect a compiler to be able to determine that.  

The traceback I received was sufficiently long that I missed the crucial piece of information that would have led me to a solution.  I resolved my issue by using pdb and discovering which symbol was missing (`func`).  After looking through the traceback again I now see that this *was* found and nicely reported, but that a second error ended up extending the traceback significantly, hiding the information I needed.

Anyway, feel free to close.

```python
import numpy as np
from numba import jit

@jit(nogil=True)
def loglike(Xbeta, y):
    eXbeta = np.exp(Xbeta)
    return np.sum(np.log1p(eXbeta)) - np.dot(y, Xbeta)


@jit(nogil=True)
def compute_stepsize(beta, step, Xbeta, Xstep, y, curr_val, stepSize=1.0,
        armijoMult=0.1, backtrackMult=0.1):
    obeta, oXbeta = beta, Xbeta
    steplen = (step**2).sum()
    lf = curr_val
    # func = 0  #  <<<---- Adding this line fixes everything
    for ii in range(100):
        beta = obeta - stepSize * step
        if ii and np.array_equal(beta, obeta):
            stepSize = 0
            break
        Xbeta = oXbeta - stepSize * Xstep

        func = loglike(Xbeta, y)
        df = lf - func
        if df >= armijoMult * stepSize * steplen:
            break
        stepSize *= backtrackMult

    return stepSize, beta, Xbeta, func

x = np.ones(5)
compute_stepsize(x, x, x, x, x, 1)
```

### Traceback

```python
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/errors.py in new_error_context(fmt_, *args, **kwargs)
    242     try:
--> 243         yield
    244     except NumbaError as e:

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/lowering.py in lower_block(self, block)
    196                                    loc=self.loc, errcls_=defaulterrcls):
--> 197                 self.lower_inst(inst)
    198 

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/objmode.py in lower_inst(self, inst)
     64         if isinstance(inst, ir.Assign):
---> 65             value = self.lower_assign(inst)
     66             self.storevar(value, inst.target.name)

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/objmode.py in lower_assign(self, inst)
    158         elif isinstance(value, ir.Expr):
--> 159             return self.lower_expr(value)
    160         elif isinstance(value, ir.Global):

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/objmode.py in lower_expr(self, expr)
    241         elif expr.op == 'call':
--> 242             argvals = [self.loadvar(a.name) for a in expr.args]
    243             fn = self.loadvar(expr.func.name)

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/objmode.py in <listcomp>(.0)
    241         elif expr.op == 'call':
--> 242             argvals = [self.loadvar(a.name) for a in expr.args]
    243             fn = self.loadvar(expr.func.name)

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/objmode.py in loadvar(self, name)
    519         # If this raises then the live variables analysis is wrong
--> 520         assert name in self._live_vars, name
    521         ptr = self.varmap[name]

AssertionError: func

During handling of the above exception, another exception occurred:

LoweringError                             Traceback (most recent call last)
<ipython-input-3-c019a5b1f546> in <module>()
     28 
     29 x = np.ones(5)
---> 30 compute_stepsize(x, x, x, x, x, 1)

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/dispatcher.py in _compile_for_args(self, *args, **kws)
    284                 argtypes.append(self.typeof_pyval(a))
    285         try:
--> 286             return self.compile(tuple(argtypes))
    287         except errors.TypingError as e:
    288             # Intercept typing error that may be due to an argument

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/dispatcher.py in compile(self, sig)
    530 
    531             self._cache_misses[sig] += 1
--> 532             cres = self._compiler.compile(args, return_type)
    533             self.add_overload(cres)
    534             self._cache.save_overload(sig, cres)

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/dispatcher.py in compile(self, args, return_type)
     79                                       impl,
     80                                       args=args, return_type=return_type,
---> 81                                       flags=flags, locals=self.locals)
     82         # Check typing error if object mode is used
     83         if cres.typing_error is not None and not flags.enable_pyobject:

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/compiler.py in compile_extra(typingctx, targetctx, func, args, return_type, flags, locals, library)
    691     pipeline = Pipeline(typingctx, targetctx, library,
    692                         args, return_type, flags, locals)
--> 693     return pipeline.compile_extra(func)
    694 
    695 

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/compiler.py in compile_extra(self, func)
    348         self.lifted = ()
    349         self.lifted_from = None
--> 350         return self._compile_bytecode()
    351 
    352     def compile_ir(self, func_ir, lifted=(), lifted_from=None):

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/compiler.py in _compile_bytecode(self)
    656         """
    657         assert self.func_ir is None
--> 658         return self._compile_core()
    659 
    660     def _compile_ir(self):

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/compiler.py in _compile_core(self)
    643 
    644         pm.finalize()
--> 645         res = pm.run(self.status)
    646         if res is not None:
    647             # Early pipeline completion

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/compiler.py in run(self, status)
    234                     # No more fallback pipelines?
    235                     if is_final_pipeline:
--> 236                         raise patched_exception
    237                     # Go to next fallback pipeline
    238                     else:

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/compiler.py in run(self, status)
    226                 try:
    227                     event(stage_name)
--> 228                     stage()
    229                 except _EarlyPipelineCompletion as e:
    230                     return e.result

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/compiler.py in stage_objectmode_frontend(self)
    419         if self.flags.enable_looplift:
    420             assert not self.lifted
--> 421             cres = self.frontend_looplift()
    422             if cres is not None:
    423                 raise _EarlyPipelineCompletion(cres)

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/compiler.py in frontend_looplift(self)
    409                               self.args, self.return_type,
    410                               outer_flags, self.locals,
--> 411                               lifted=tuple(loops), lifted_from=None)
    412             return cres
    413 

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/compiler.py in compile_ir(typingctx, targetctx, func_ir, args, return_type, flags, locals, lifted, lifted_from, library)
    705                         args, return_type, flags, locals)
    706     return pipeline.compile_ir(func_ir=func_ir, lifted=lifted,
--> 707                                lifted_from=lifted_from)
    708 
    709 

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/compiler.py in compile_ir(self, func_ir, lifted, lifted_from)
    356 
    357         self._set_and_check_ir(func_ir)
--> 358         return self._compile_ir()
    359 
    360     def stage_analyze_bytecode(self):

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/compiler.py in _compile_ir(self)
    663         """
    664         assert self.func_ir is not None
--> 665         return self._compile_core()
    666 
    667 

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/compiler.py in _compile_core(self)
    643 
    644         pm.finalize()
--> 645         res = pm.run(self.status)
    646         if res is not None:
    647             # Early pipeline completion

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/compiler.py in run(self, status)
    234                     # No more fallback pipelines?
    235                     if is_final_pipeline:
--> 236                         raise patched_exception
    237                     # Go to next fallback pipeline
    238                     else:

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/compiler.py in run(self, status)
    226                 try:
    227                     event(stage_name)
--> 228                     stage()
    229                 except _EarlyPipelineCompletion as e:
    230                     return e.result

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/compiler.py in stage_objectmode_backend(self)
    559         """
    560         lowerfn = self.backend_object_mode
--> 561         self._backend(lowerfn, objectmode=True)
    562 
    563         # Warn if compiled function in object mode and force_pyobject not set

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/compiler.py in _backend(self, lowerfn, objectmode)
    536             self.library.enable_object_caching()
    537 
--> 538         lowered = lowerfn()
    539         signature = typing.signature(self.return_type, *self.args)
    540         self.cr = compile_result(typing_context=self.typingctx,

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/compiler.py in backend_object_mode(self)
    510                                      self.library,
    511                                      self.func_ir,
--> 512                                      self.flags)
    513 
    514     def backend_nopython_mode(self):

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/compiler.py in py_lowering_stage(targetctx, library, interp, flags)
    833     fndesc = funcdesc.PythonFunctionDescriptor.from_object_mode_function(interp)
    834     lower = objmode.PyLower(targetctx, library, fndesc, interp)
--> 835     lower.lower()
    836     if not flags.no_cpython_wrapper:
    837         lower.create_cpython_wrapper()

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/lowering.py in lower(self)
    120         if self.generator_info is None:
    121             self.genlower = None
--> 122             self.lower_normal_function(self.fndesc)
    123         else:
    124             self.genlower = self.GeneratorLower(self)

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/lowering.py in lower_normal_function(self, fndesc)
    155         # Init argument values
    156         self.extract_function_arguments()
--> 157         entry_block_tail = self.lower_function_body()
    158 
    159         # Close tail of entry block

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/lowering.py in lower_function_body(self)
    180             bb = self.blkmap[offset]
    181             self.builder.position_at_end(bb)
--> 182             self.lower_block(block)
    183 
    184         self.post_lower()

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/lowering.py in lower_block(self, block)
    195             with new_error_context('lowering "{inst}" at {loc}', inst=inst,
    196                                    loc=self.loc, errcls_=defaulterrcls):
--> 197                 self.lower_inst(inst)
    198 
    199     def create_cpython_wrapper(self, release_gil=False):

/home/mrocklin/Software/anaconda/lib/python3.5/contextlib.py in __exit__(self, type, value, traceback)
     75                 value = type()
     76             try:
---> 77                 self.gen.throw(type, value, traceback)
     78                 raise RuntimeError("generator didn't stop after throw()")
     79             except StopIteration as exc:

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/errors.py in new_error_context(fmt_, *args, **kwargs)
    247     except Exception as e:
    248         newerr = errcls(e).add_context(_format_msg(fmt_, args, kwargs))
--> 249         six.reraise(type(newerr), newerr, sys.exc_info()[2])
    250 
    251 

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/six.py in reraise(tp, value, tb)
    656             value = tp()
    657         if value.__traceback__ is not tb:
--> 658             raise value.with_traceback(tb)
    659         raise value
    660 

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/errors.py in new_error_context(fmt_, *args, **kwargs)
    241     errcls = kwargs.pop('errcls_', InternalError)
    242     try:
--> 243         yield
    244     except NumbaError as e:
    245         e.add_context(_format_msg(fmt_, args, kwargs))

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/lowering.py in lower_block(self, block)
    195             with new_error_context('lowering "{inst}" at {loc}', inst=inst,
    196                                    loc=self.loc, errcls_=defaulterrcls):
--> 197                 self.lower_inst(inst)
    198 
    199     def create_cpython_wrapper(self, release_gil=False):

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/objmode.py in lower_inst(self, inst)
     63     def lower_inst(self, inst):
     64         if isinstance(inst, ir.Assign):
---> 65             value = self.lower_assign(inst)
     66             self.storevar(value, inst.target.name)
     67 

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/objmode.py in lower_assign(self, inst)
    157             return val
    158         elif isinstance(value, ir.Expr):
--> 159             return self.lower_expr(value)
    160         elif isinstance(value, ir.Global):
    161             return self.lower_global(value.name, value.value)

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/objmode.py in lower_expr(self, expr)
    240             return res
    241         elif expr.op == 'call':
--> 242             argvals = [self.loadvar(a.name) for a in expr.args]
    243             fn = self.loadvar(expr.func.name)
    244             args = self.pyapi.tuple_pack(argvals)

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/objmode.py in <listcomp>(.0)
    240             return res
    241         elif expr.op == 'call':
--> 242             argvals = [self.loadvar(a.name) for a in expr.args]
    243             fn = self.loadvar(expr.func.name)
    244             args = self.pyapi.tuple_pack(argvals)

/home/mrocklin/Software/anaconda/lib/python3.5/site-packages/numba/objmode.py in loadvar(self, name)
    518         """
    519         # If this raises then the live variables analysis is wrong
--> 520         assert name in self._live_vars, name
    521         ptr = self.varmap[name]
    522         val = self.builder.load(ptr)

LoweringError: Failed at object (object mode frontend)
Failed at object (object mode backend)
func
File "<ipython-input-3-c019a5b1f546>", line 14
[1] During: lowering "$75 = call $74(Xbeta, Xstep, armijoMult, backtrackMult, beta, func, lf, oXbeta, obeta, step, stepSize, steplen, y)" at <ipython-input-3-c019a5b1f546> (14)
```

**Comment by sklam** [comment]
Thanks for the report.   This is indeed an internal error as that assertion is not suppose to raise.  It raising indicates a problem at the live variable analysis.  This is an interesting case for us to study.

**Comment by stuartarchibald** [comment]
Once merged https://github.com/numba/numba/pull/3230 fixes this and the giant tracebacks ought to only appear in the case there's an internal error following on from work under https://github.com/numba/numba/issues/2888.

**Comment by seibert** [comment]
Closed by #3230.
